### ISSUE TITLE
Add support for ADT enums in query/path params

### ISSUE THREAD

**Issue by matwojcik** [softwaremill/tapir#200]
It would be nice to have some implicit codec for ADT enums, e.g.:

```scala
sealed trait Status extends Product with Serializable
case object Initial extends Status
case object Finished extends Status
```

would enable building routes `/path/to/:status` basing on Status name, e.g. `/path/to/Initial`, `/path/to/Finished`

**Comment by adamw** [comment]
Maybe through integration w/ https://www.github.com/lloydmeta/enumeratum ?

**Comment by matwojcik** [comment]
Haven't used https://github.com/lloydmeta/enumeratum#inheritance-free-usage but it seems like it could be used, to omit the need to inherit from `Enum`.

Would you include this dependency on `enumeratum` in `core` or some module?

**Comment by adamw** [comment]
As a module. Unless you can somehow support ADT-enums in another way, but I can't think of a generic way right now :)

**Comment by skennedy** [comment]
Have just written some instances for enumeratum integration in my own project. 

@adamw would be delighted to contribute back as a new module. Would you accept a PR for this?

**Comment by adamw** [comment]
@skennedy Sure! There's one for cats already, which paved the way :)

**Comment by skennedy** [comment]
Cool, I'll use it as inspiration

**Comment by adamw** [comment]
I was trying to come up with a name which could group those two modules ... sth like `codec-integrations`?

**Comment by PawelJ-PL** [comment]
Default implementation generates invalid schema for `EnumEntry`. Following class:

```scala
final case class FirstDto(foo: String, bar: Long, baz: Option[Boolean], dataType: DataType)

object FirstDto {
  implicit val encoder: Encoder[FirstDto] = deriveEncoder[FirstDto]
  implicit val decoder: Decoder[FirstDto] = deriveDecoder[FirstDto]
}

sealed trait DataType extends EnumEntry

object DataType extends Enum[DataType] with CirceEnum[DataType] {
  val values = findValues

  case object Some extends DataType
  case object Other extends DataType
}

```

will generate OpenApi doc:

```yaml
FirstDto:
      required:
      - foo
      - bar
      - dataType
      type: object
      properties:
        foo:
          type: string
        bar:
          type: integer
          format: int64
        baz:
          type: boolean
        dataType:
          $ref: '#/components/schemas/DataType'
    DataType:
      oneOf:
      - $ref: '#/components/schemas/Other'
      - $ref: '#/components/schemas/Some'
    Other:
      type: object
    Some:
      type: object
```

and I think it should be something like:

```yaml
FirstDto:
      required:
      - foo
      - bar
      - dataType
      type: object
      properties:
        foo:
          type: string
        bar:
          type: integer
          format: int64
        baz:
          type: boolean
        dataType:
          $ref: '#/components/schemas/DataType'
    DataType:
      oneOf:
      - $ref: '#/components/schemas/Other'
      - $ref: '#/components/schemas/Some'
    Other:
      type: string
    Some:
      type: string
```

I can add
```scala
implicit val schemaForEnum: Schema[DataType] = Schema(SchemaType.SString)
```

But it will cause, that `DataType` will be interpreted as any string, so doc will not include information about possible values.

@skennedy did you deal with this problem?

**Comment by skennedy** [comment]
@PawelJ-PL you are right to define the Schema like that, but I think you are missing a `Validator[DataType]` instance - as that defines the possible values of your data type. 

So I think adding this along with your `Schema` should fix it:
```scala
implicit val validatorForEnum: Validator[DataType] = Validator.enum(DataType.values.toList, v => Some(v.entryName)
```

I have a WIP branch to add common support, which I hope to raise PR for this week: 
https://github.com/softwaremill/tapir/compare/master...skennedy:enumeratum

**Comment by PawelJ-PL** [comment]
@skennedy thank you. With `Validator` instance it generates:

```yaml
FirstDto:
      required:
      - foo
      - bar
      - dataType
      type: object
      properties:
        foo:
          type: string
        bar:
          type: integer
          format: int64
        baz:
          type: boolean
        dataType:
          type: string
          enum:
          - Some
          - Other
```

I wasn't aware, that I need custom `Validator` for output.

**Comment by adamw** [comment]
@PawelJ-PL the enumeraturm integration by @skennedy is released in 0.12.20, maybe this will help you :)

**Comment by PawelJ-PL** [comment]
I've upgraded Tapir to `0.12.20` - migration from `0.11.9` was not easy ;) and it looks, that `tapir-enumeratum` works as expected. Thank you :)
